### PR TITLE
[AWS SDK] Additional aws encoding details for the Rest Json protocol 

### PR DESCRIPTION
**Pull request by Baccata**
Besides the MD5 test (which appears to be false-positive), this PR takes care of tweaking the last bits that need tweaking for the AWS Rest Json protocol to be green. 

Depends on https://github.com/disneystreaming/smithy4s/pull/1035

(https://github.com/disneystreaming/smithy4s/pull/1035 should be merged before this one)

### CODE REVIEW

File: modules/aws-http4s/src/smithy4s/aws/json/AwsJsonCodecAPI.scala [hunk 1]
```diff
@@ -16,8 +16,27 @@
...
+    hintMask: Option[HintMask] = None
+) extends smithy4s.http.json.JsonCodecAPI(
       new AwsSchemaVisitorJCodec(_),
-      None
-    ) {}
+      hintMask
+    ) {
+
+  override def writeToArray[A](codec: Codec[A], value: A): Array[Byte] = {
+    // AWS expects an empty object to be sent by clients even when the data
```
Comment (daddykotex) [inline modules/aws-http4s/src/smithy4s/aws/json/AwsJsonCodecAPI.scala:32]: is this related to the changes I've made in: https://github.com/disneystreaming/smithy4s/pull/1022/files#diff-3d68306dcc7da993cfe2a9162443c1ff9993076618cc20aaeaa58e8244753c8cL90

after the changes in the alloy spec: https://github.com/disneystreaming/alloy/pull/96/files

Comment (Baccata) [reply modules/aws-http4s/src/smithy4s/aws/json/AwsJsonCodecAPI.scala:32]: It is, but your change targeted the 0.17 branch. Everything has changed in the meantime  :)

File: modules/aws-http4s/src/smithy4s/aws/json/AwsSchemaVisitorJCodec.scala [hunk 2]
```diff
@@ -105,6 +107,84 @@ private[aws] class AwsSchemaVisitorJCodec(cache: CompilationCache[JCodec])
...
+            if (in.isCurrentToken('}')) builder.result()
+            else in.objectEndOrCommaError()
+          }
+        } else in.decodeError("Expected JSON object")
+
+      def encodeValue(xs: Map[K, V], out: JsonWriter): Unit = {
+        out.writeObjectStart()
+        xs.foreach { kv =>
+          jk.encodeKey(kv._1, out)
+          jv.encodeValue(kv._2, out)
```
Comment (daddykotex) [inline modules/aws-http4s/src/smithy4s/aws/json/AwsSchemaVisitorJCodec.scala:172]: do you have to encode the comma?

Comment (daddykotex) [reply modules/aws-http4s/src/smithy4s/aws/json/AwsSchemaVisitorJCodec.scala:172]: I'm assuming we don't or we would have some tests blow up

**Review by lewisjkl — APPROVED** [review]

File: modules/aws-http4s/src/smithy4s/aws/json/AwsSchemaVisitorJCodec.scala (hunk 2 above)
Comment (Baccata) [reply modules/aws-http4s/src/smithy4s/aws/json/AwsSchemaVisitorJCodec.scala:172]: Yeah, `encodeKey` takes care of it, as it delegates to the `writeKey` methods, which look like this in jsoniter : 

``` 
  def writeKey(x: Int): Unit = {
    writeOptionalCommaAndIndentionBeforeKey()
    writeBytes('"')
    writeInt(x)
    writeParenthesesWithColon()
  }
```

**Pull request merged**